crates/node-builder/src/engine.rs

```diff
@@ -1,23 +1,58 @@
-use reth_node_api::EngineTypes;
-use reth_payload_builder::{EthPayloadBuilderAttributes, OptimismPayloadBuilderAttributes};
-use reth_rpc_types::engine::{OptimismPayloadAttributes, PayloadAttributes};
+#[cfg(feature = "optimism")]
+use reth_node_api::optimism_validate_version_specific_fields;
+use reth_node_api::{
+    validate_version_specific_fields, AttributesValidationError, EngineApiMessageVersion,
+    EngineTypes, PayloadAttributes, PayloadOrAttributes,
+};
+use reth_payload_builder::EthPayloadBuilderAttributes;
+#[cfg(feature = "optimism")]
+use reth_payload_builder::OptimismPayloadBuilderAttributes;
+use reth_primitives::ChainSpec;
+#[cfg(feature = "optimism")]
+use reth_rpc_types::engine::OptimismPayloadAttributes;
+use reth_rpc_types::engine::PayloadAttributes as EthPayloadAttributes;
 
 /// The types used in the default mainnet ethereum beacon consensus engine.
 #[derive(Debug, Default, Clone)]
 #[non_exhaustive]
 pub struct EthEngineTypes;
 
 impl EngineTypes for EthEngineTypes {
-    type PayloadAttributes = PayloadAttributes;
+    type PayloadAttributes = EthPayloadAttributes;
     type PayloadBuilderAttributes = EthPayloadBuilderAttributes;
+
+    fn validate_version_specific_fields<Type>(
+        chain_spec: &ChainSpec,
+        version: EngineApiMessageVersion,
+        payload_or_attrs: &PayloadOrAttributes<'_, Type>,
+    ) -> Result<(), AttributesValidationError>
+    where
+        Type: PayloadAttributes,
+    {
+        validate_version_specific_fields(chain_spec, version, payload_or_attrs)
+    }
 }
 
+#[cfg(feature = "optimism")]
 /// The types used in the optimism beacon consensus engine.
 #[derive(Debug, Default, Clone)]
 #[non_exhaustive]
 pub struct OptimismEngineTypes;
 
+// TODO: remove cfg once Hardfork::Canyon can be used without the flag
+#[cfg(feature = "optimism")]
 impl EngineTypes for OptimismEngineTypes {
     type PayloadAttributes = OptimismPayloadAttributes;
     type PayloadBuilderAttributes = OptimismPayloadBuilderAttributes;
+
+    fn validate_version_specific_fields<Type>(
+        chain_spec: &ChainSpec,
+        version: EngineApiMessageVersion,
+        payload_or_attrs: &PayloadOrAttributes<'_, Type>,
```

Comment:
this doesn't need to be a reference because this type already only contains references.